Adding fall back method for `sha1` in case default algo is `sha256`

awais786 opened the pull request:
https://docs.djangoproject.com/en/3.2/ref/settings/#default-hashing-algorithm

https://docs.djangoproject.com/en/4.2/releases/4.0/#features-removed-in-4-0
The `DEFAULT_HASHING_ALGORITHM` transitional setting is removed.


~~its defined in django-global settings as 256
https://github.com/django/django/blob/stable/3.2.x/django/conf/global_settings.py#L450~~

~~Initially we were assuming  that platform is going to use `sha256` as default algorithm for every thing. But now we got clarity that `safesessionmiddleware` will use the `sha1` so created new [PR](https://github.com/openedx/edx-platform/pull/33380) with new variable.~~


~~**Testing steps**~~
~~1. create sandbox with master branch and login on sandbox. 
2. install `django42` and restart `lms` only.
3. keep continue using the site, some `middleware` call that [method](https://github.com/django/django/blob/4.2.5/django/contrib/auth/base_user.py#L151). It will mismatch due to algorithm and logouts the user.
4. Also moving to `django42` will invalidates the forgot passwords links generated via `django32`.~~


Picking `SHA256` value from internal config for stage. https://github.com/edx/edx-internal/pull/9332
For production it will remain `sha1`.

awais786 committed:
feat!: `sha1` has been deprecated in django32 and removed in django42.

jmbowman commented:
It looks like most/all of the `sha1` instances are imported from the standard library via `hashlib`.  We may want to revisit those to see if another hash is more appropriate, but that shouldn't block the Django upgrade.

robrap commented on openedx/core/djangoapps/cache_toolbox/middleware.py:
```diff
@@ -141,9 +142,21 @@ def _verify_session_auth(self, request):
+            session_hash_verified = session_hash and constant_time_compare(
+                session_hash, request.user.get_session_auth_hash())
+            if session_hash_verified:
+                set_custom_attribute('CacheBackedAuthenticationMiddleware.verify.alg', settings.DEFAULT_HASHING_ALGORITHM)
```
Minor suggestions:
1. Monitoring attributes will always be set, and aren't in a special case, so you'll have all the data you might be wondering about.
2. Returning early you can keep or ignore, but just allows for fewer if/else and might make the code a little more readable.
3. If you find a different place for `settings.DEFAULT_HASHING_ALGORITHM` custom attribute, it doesn't need to be repeated. See other PR comments.
```suggestion
            # temporary monitoring for transition to new hashing algorithm
            set_custom_attribute('DEFAULT_HASHING_ALGORITHM', settings.DEFAULT_HASHING_ALGORITHM)
            set_custom_attribute('session_hash_verified', session_hash_verified)
            if session_hash_verified:
                return
```

robrap commented on openedx/core/djangoapps/cache_toolbox/middleware.py:
```diff
@@ -141,9 +142,21 @@ def _verify_session_auth(self, request):
+                    request.user = AnonymousUser()
+                    _mark_cookie_for_deletion(request)
+                else:
+                    set_custom_attribute('CacheBackedAuthenticationMiddleware.verify.alg', "sha1")
```
Nit: 
1. You can decide on this one, but I like keeping my monitoring with the code, rather than creating additional cases. That said, you can do what you wish here.
2. Not sure if it would be useful to have the result of `hasattr(request.user, '_legacy_get_session_auth_hash')` in a custom attribute?
```suggestion
                    # temporary monitoring for failed legacy session check
                    set_custom_attribute('failed_legacy_session_verification', True)
```

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -272,6 +272,8 @@ def sign(self, user_id):
             salt=self.key_salt, algorithm=settings.DEFAULT_HASHING_ALGORITHM
         ).sign_object(data_to_sign, serializer=signing.JSONSerializer, compress=False)
 
+        set_custom_attribute('SafeCookieData.verify.alg', settings.DEFAULT_HASHING_ALGORITHM)
```
You only need to set `settings.DEFAULT_HASHING_ALGORITHM` in one place. I'd recommend using `DEFAULT_HASHING_ALGORITHM` as the attribute name, and doing it near the top of the most relevant middleware.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -284,7 +286,18 @@ def verify(self, user_id):
             ).unsign_object(self.signature, serializer=signing.JSONSerializer, max_age=settings.SESSION_COOKIE_AGE)
 
             if unsigned_data == self._compute_digest(user_id):
+                set_custom_attribute('SafeCookieData.verify.alg', settings.DEFAULT_HASHING_ALGORITHM)
```
Nit: But this will make it more clear what line of code is running. You should have `settings.DEFAULT_HASHING_ALGORITHM` in another attribute, so you won't be losing data.
Also, I changed `verify` to `verified`, because this was actually verified, not just used to verify.
```suggestion
                set_custom_attribute('SafeCookieData.verified.alg', 'default')
```

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -284,7 +286,18 @@ def verify(self, user_id):
+                ).unsign_object(self.signature, serializer=signing.JSONSerializer, max_age=settings.SESSION_COOKIE_AGE)
+
+                if unsigned_data == self._compute_digest(user_id):
+                    set_custom_attribute('SafeCookieData.verify.alg', "sha1")
```
Nit: If we switch the above value to `default`, it doesn't really matter which we use here. So, just something to consider to make it clear what line is being executed.
Again, I also changed `verify` to `verified`.
```suggestion
                    set_custom_attribute('SafeCookieData.verified.alg', "fallback")
```

robrap commented on cms/envs/common.py:
```diff
@@ -1085,6 +1085,8 @@
 
 DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'
+
+# This will be overridden through LMS_CFG 
```
LMS or CMS?

robrap reviewed (commented):
A bunch of minor comments around monitoring.

iamsobanjaved commented:
@robrap Just fixing tests, will then implement your suggestions regarding monitoring

robrap commented on openedx/core/djangoapps/cache_toolbox/middleware.py:
```diff
@@ -144,24 +145,27 @@ def _verify_session_auth(self, request):
-                    set_custom_attribute('CacheBackedAuthenticationMiddleware.verify.alg', "sha1")
+            ):
+                # session hash is verified from legacy hashing algorithm.
+                set_custom_attribute('legacy_session_hash_verified', True)
```
[Nit] Might be simpler to use of you rename this to `session_hash_verified` and make the value "fallback" (or "legacy"), and change it to "default" (from True) above.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -296,7 +294,7 @@ def verify(self, user_id):
 
                 if unsigned_data == self._compute_digest(user_id):
-                    set_custom_attribute('SafeCookieData.verify.alg', "sha1")
+                    set_custom_attribute('SafeCookieData.verified.alg', "fallabck")
```
```suggestion
                    set_custom_attribute('SafeCookieData.verified.alg', "fallback")
```

robrap commented on openedx/core/djangoapps/cache_toolbox/tests/test_middleware.py:
```diff
@@ -37,22 +39,26 @@ def _test_change_session_hash(self, test_url, redirect_url, target_status_code=2
         self._test_change_session_hash(dashboard_url, reverse('signin_user') + '?next=' + dashboard_url)
 
     @skip_unless_cms
+    @pytest.mark.skipif(django.VERSION[0] >= 4, reason="method `_legacy_get_session_auth_hash` removed in django42")
```
Is there a Django 4 test that makes sure that all works fine without `_legacy_get_session_auth_hash`?

awais786 replied:
Added new test to run on django42.

robrap reviewed (commented):
Looks good. Minor comments. You don't need to wait for another round from me before approving/merging when you are ready.

robrap commented on openedx/core/djangoapps/safe_sessions/middleware.py:
```diff
@@ -284,7 +284,18 @@ def verify(self, user_id):
+                ).unsign_object(self.signature, serializer=signing.JSONSerializer, max_age=settings.SESSION_COOKIE_AGE)
+
+                if unsigned_data == self._compute_digest(user_id):
+                    set_custom_attribute('SafeCookieData.verified.alg', "fallback")
```
Consider adding a "Temporary Hash Change Rollout" dashboard in New Relic with charts like:
```
SELECT count(*) FROM Transaction WHERE SafeCookieData.verified.alg IS NOT NULL FACET SafeCookieData.verified.alg TIMESERIES MAX
```
Note: I'm not sure if New Relic will require quotes with the attribute name with `.`, so you could also rename it to `SafeCookieData_verified_alg`, or you may need to use surrounding quotes like ` in the NRQL.

The dashboard could also include logins/logouts with `response.status`, or anything else you'd want to track to ensure all is well and they we won't want to temporarily rollback any config change.

robrap commented on openedx/core/djangoapps/cache_toolbox/tests/test_middleware.py:
```diff
@@ -37,8 +40,9 @@ def _test_change_session_hash(self, test_url, redirect_url, target_status_code=2
-            self.assertRedirects(response, redirect_url, target_status_code=target_status_code)
+            with patch.object(User, '_legacy_get_session_auth_hash', return_value='abc123'):
+                response = self.client.get(test_url)
+                self.assertRedirects(response, redirect_url, target_status_code=target_status_code)
```
Is this patching the only difference between the Django version tests? Could something like the following get rid of most of the test duplication?
```suggestion
            # Django 3.2 has _legacy_get_session_auth_hash, and Django 4 does not
            if hasattr(User, '_legacy_get_session_auth_hash'):  # not sure if this works, or something like this????
                with patch.object(User, '_legacy_get_session_auth_hash', return_value='abc123'):
                    response = self.client.get(test_url)
            else:
                response = self.client.get(test_url)

            self.assertRedirects(response, redirect_url, target_status_code=target_status_code)
```

Also, maybe adding `autospec=True` to the first `patch.object` will also help protect you more?

iamsobanjaved replied:
Updated

robrap reviewed (commented):
I think it is ok, but it is hard to review with all the test duplication, and not knowing what is changing in each case. I made a potential suggestion.

robrap commented on openedx/core/djangoapps/cache_toolbox/middleware.py:
```diff
@@ -141,9 +143,29 @@ def _verify_session_auth(self, request):
+
+            # session hash is verified from the default algo, so skip legacy check
+            if session_hash_verified:
+                set_custom_attribute('session_hash_verified', "default")
```
Quick note that adding test assertions for these custom attributes can be a great way to prove to yourself you’ve got all the coverage you want from your tests.

iamsobanjaved replied:
Added

Commit:
test: fix quality failure

robrap commented on openedx/core/djangoapps/cache_toolbox/tests/test_middleware.py:
```diff
@@ -36,10 +37,33 @@ def _test_change_session_hash(self, test_url, redirect_url, target_status_code=2
 
+    def _test_custom_attribute_after_changing_hash(self, test_url, mock_set_custom_attribute):
+        """verify that set_custom_attribute is called with expected values"""
+        response = self.client.get(test_url)
```
I had some difficulty knowing the cases being tested. I was trying to write a comment, but it got out of hand, so instead I created a PR: https://github.com/openedx/edx-platform/pull/33436. Note that there are some other minor fixes in the PR as well.

robrap reviewed (commented):
I think this is probably all set with my PR. I just wanted to ensure that all cases were being covered.

robrap committed:
fixup! update custom attribute tests (#33436)

I was wondering about all the cases, so I
updated the test to reflect this. I also
made some other minor adjustments.

iamsobanjaved committed:
Merge branch 'master' into removing-sha1

robrap reviewed (approved):
Thanks so much for this.

A possible deployment plan:
1. Merge
2. Create (and share) a small New Relic dashboard (ask if you need help). If you FACET by appName (or add separate charts for Stage), we can watch in Stage for a bit before going to Prod.
3. Enable in Stage and ensure all is well. (Maybe leaves this for a day?)
4. Enable in Prod.
Thoughts?

iamsobanjaved commented:
Yeah, this is what we had in our minds for deployment. I will try to create a New Relic dashboard and surely ask for help if I will get stuck. Also merging this PR now.
